### PR TITLE
Allow users to provide additional Custom Resources

### LINKED ISSUE #130: Allow users to run custom resource yaml manifests from a flag for improved CI automation
In order to help developers run their full setup with opcap as the tester we could allow for complicated CRs that are not included on the CSV's  ALMExamples field to be injected from an external file providing the path through a flag. And possibly giving the opportunity to bypass the ALMExamples if it's intentional. That opens the door to a whole new level of CI automation.

### PR DESCRIPTION
**Pull request by mgoerens**
## Description of PR

Adds the '--extra-cr-directory' flag to the check command. The expected directory structures is:
```
custom_resources_directory/
├── package_name1
│   ├── manifest_file1.json
│   └── manifest_file2.yaml
└── package_name2
    ├── manifest_file1.json
    └── manifest_file2.yaml
```
Additional CRs can be specified for multiple packages by creating multiple subdirectories. The name of the subdirectory should correspond to the name of the package the manifest files are to be be deployed with.

Multiple manifest files can be specified for each package. Both JSON and YAML formats are accepted.

The additional CRs are merged with the ALMExamples, if any.

Fixes #130

## Changes (required)

- Add the `--extra-cr-directory` flag to the check command
- Load extra CRs from this directory.
- Add a `ExtraCustomResources` field to the `capAuditor` struct, associating packages to a list of extra CRs.
- Add the extra CR to corresponding Audits.
- Edit the `extractAlmExamples` function, so that both ALMExamples and extra CRs are installed.

## Checklist (required)
- [x] I have reviewed and followed the [contribution guidelines](https://github.com/opdev/opcap/blob/main/docs/contribution.md)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if needed
- [x] I have checked that my changes pass all tests

## References (optional)

### CODE REVIEW

File: cmd/check.go [hunk 1]
```diff
@@ -21,10 +26,74 @@ type CheckCommandFlags struct {
...
+	}
+
+	for _, packageDirectory := range extraCRDirectoryContent {
+		if packageDirectory.IsDir() {
+			packageName := packageDirectory.Name()
+			packageDirectoryPath := filepath.Join(extraCRDirectoryAbsolutePath, packageName)
+			createPackageKey := false
+			customResourceManifests := []map[string]interface{}{}
+
+			packageDirectoryContent, err := ioutil.ReadDir(packageDirectoryPath)
```
Comment (bcrochet) [inline cmd/check.go:55]: ioutil is deprecated. os.ReadDir should be used instead.

File: cmd/check.go [hunk 2]
```diff
@@ -21,10 +26,74 @@ type CheckCommandFlags struct {
...
+			}
+
+			for _, manifestFile := range packageDirectoryContent {
+				if manifestFile.Mode().IsRegular() {
+					manifestFilePath := filepath.Join(packageDirectoryPath, manifestFile.Name())
+					createPackageKey = true
+
+					logger.Debugw("adding Custom Resource", "source manifest file", manifestFilePath, "package", packageName)
+
+					manifestFileContentRaw, err := ioutil.ReadFile(manifestFilePath)
```
Comment (bcrochet) [inline cmd/check.go:67]: ioutil is deprecated. Use os.ReadFile.

File: cmd/check.go [hunk 3]
```diff
@@ -73,6 +142,10 @@ and/or users.`,
 				AllInstallModes:        checkflags.AllInstallModes,
 			}
 
+			if checkflags.ExtraCRDirectory != "" {
+				handleExtraCRDirectoryFlag(capAuditor)
```
Comment (bcrochet) [inline cmd/check.go:146]: Should this instead be a part of capAuditor? 

```
capAuditor.ExtraCRDirectory(checkflags.ExtraCRDirectory)
```

File: cmd/check.go [hunk 4]
```diff
@@ -89,6 +162,8 @@ and/or users.`,
 	flags.StringSliceVar(&checkflags.AuditPlan, "audit-plan", defaultAuditPlan, "audit plan is the ordered list of operator test functions to be called during a capability audit.")
 	flags.StringSliceVar(&checkflags.Packages, "packages", []string{}, "a list of package(s) which limits audits and/or other flag(s) output")
 	flags.BoolVar(&checkflags.AllInstallModes, "all-installmodes", false, "when set, all install modes supported by an operator will be tested")
+	flags.StringVar(&checkflags.ExtraCRDirectory, "extra-cr-directory", "",
+		"directory containing the additional Custom Resources to be deployed by the OperandInstall audit. The manifest files should be located in subdirectories named after the packages they are corresponding to.")
```
Comment (bcrochet) [inline cmd/check.go:166]: Can this be specified multiple times?

**Review by bcrochet — CHANGES_REQUESTED** [review]
⚠️  Suggestions and questions.

File: cmd/check.go (hunk 4 above)
Comment (mgoerens) [reply cmd/check.go:166]: No it should not. You specify for instance `--extra-cr-directory custom_resources_directory/` and the following structure is expected:
```
custom_resources_directory/
├── package_name1
│   ├── manifest_file1.json
│   └── manifest_file2.yaml
└── package_name2
    ├── manifest_file1.json
    └── manifest_file2.yaml
```

So you set only one `--extra-cr-directory` in the CLI, but you can target multiple packages.

File: cmd/check.go (hunk 1 above)
Comment (mgoerens) [reply cmd/check.go]: Fixed in eea81a145e7b39a2af3820518feb75373b791b48 (/!\ Fixup commit, don't automerge)

File: cmd/check.go (hunk 2 above)
Comment (mgoerens) [reply cmd/check.go]: Fixed in eea81a145e7b39a2af3820518feb75373b791b48 (/!\ Fixup commit, don't automerge)

File: cmd/check.go (hunk 3 above)
Comment (mgoerens) [reply cmd/check.go]: I really don't have a strong opinion here. I wasn't sure where to add this.

I followed you suggestion in ae2dfb333230741a232f7fd61edbfaf2927cdba0

**Review by bcrochet — COMMENTED** [review]
Bit of a suggestion. We can always revisit later.

File: internal/capability/auditor.go [hunk 5]
```diff
@@ -26,6 +29,79 @@ type CapAuditor struct {
...
+}
+
+// ExtraCRDirectory scans the provided directory and populates the ExtraCustomResources field
+func (capAuditor *CapAuditor) ExtraCRDirectory(extraCRDirectory string) {
+	logger.Debugw("scaning for extra Custom Resources", "extra CR directory", extraCRDirectory)
+	extraCustomResources := map[string]interface{}{} // maps packages to a list of CR
+
+	extraCRDirectoryAbsolutePath, err := filepath.Abs(extraCRDirectory)
+	if err != nil {
+		logger.Errorf("Error getting absolute path of %s: %v", extraCRDirectory, err)
```
Comment (bcrochet) [inline internal/capability/auditor.go:45]: This seems like something that should cause this to fail. You might want to change your signature to return an error.

In this case, you should return an error like:

```
return fmt.Errorf("could not get absolute path from %s: %v", extraCRDirectory, err)
```

File: internal/capability/auditor.go [hunk 6]
```diff
@@ -26,6 +29,79 @@ type CapAuditor struct {
...
+	}
+
+	extraCRDirectoryContent, err := os.ReadDir(extraCRDirectoryAbsolutePath)
+	if err != nil {
+		logger.Errorf("Couldn't read %s: %v", extraCRDirectoryAbsolutePath, err)
```
Comment (bcrochet) [inline internal/capability/auditor.go:50]: Same here. You can't continue of this fails.

```
return fmt.Errorf("could not read directory %s: %v", extraCRDirectoryAbsolutePath, err)
```

File: internal/capability/auditor.go [hunk 7]
```diff
@@ -26,6 +29,79 @@ type CapAuditor struct {
...
+	for _, packageDirectory := range extraCRDirectoryContent {
+		if packageDirectory.IsDir() {
+			packageName := packageDirectory.Name()
+			packageDirectoryPath := filepath.Join(extraCRDirectoryAbsolutePath, packageName)
+			createPackageKey := false
+			customResourceManifests := []map[string]interface{}{}
+
+			packageDirectoryContent, err := os.ReadDir(packageDirectoryPath)
+			if err != nil {
+				logger.Errorf("Error reading directory %s: %v", packageDirectoryPath, err)
```
Comment (bcrochet) [inline internal/capability/auditor.go:62]: This one you might want to just log here and then use a continue.

File: internal/capability/auditor.go [hunk 8]
```diff
@@ -26,6 +29,79 @@ type CapAuditor struct {
...
+			}
+
+			for _, manifestFile := range packageDirectoryContent {
+				manifestFileInfo, err := manifestFile.Info()
+				if err != nil {
+					logger.Errorf("Error getting info of file %s: %v", manifestFile.Name(), err)
```
Comment (bcrochet) [inline internal/capability/auditor.go:68]: Again, log and continue.

File: internal/capability/auditor.go [hunk 9]
```diff
@@ -26,6 +29,79 @@ type CapAuditor struct {
...
+
+				if manifestFileInfo.Mode().IsRegular() {
+					manifestFilePath := filepath.Join(packageDirectoryPath, manifestFile.Name())
+					createPackageKey = true
+
+					logger.Debugw("adding Custom Resource", "source manifest file", manifestFilePath, "package", packageName)
+
+					manifestFileContentRaw, err := os.ReadFile(manifestFilePath)
+					if err != nil {
+						logger.Errorf("Error reading file %s: %v", manifestFilePath, err)
```
Comment (bcrochet) [inline internal/capability/auditor.go:79]: Log and continue.

File: cmd/check.go [hunk 10]
```diff
@@ -73,6 +74,10 @@ and/or users.`,
 				AllInstallModes:        checkflags.AllInstallModes,
 			}
 
+			if checkflags.ExtraCRDirectory != "" {
+				capAuditor.ExtraCRDirectory(checkflags.ExtraCRDirectory)
```
Comment (bcrochet) [inline cmd/check.go:78]: Add an error return to ExtraCRDirectory and handle appropriately here. Should it continue if there is an error in the extras? Are they considered optional? If not, we should return an error.

**Review by bcrochet — CHANGES_REQUESTED** [review]
Clarification on errors.

File: internal/capability/auditor.go (hunk 5 above)
Comment (mgoerens) [reply internal/capability/auditor.go]: Addressed, as well as other comments regarding logging / returning error / continuing in 7bbad5d23beb1042de6f882ace95350ddad866d6

File: cmd/check.go (hunk 10 above)
Comment (mgoerens) [reply cmd/check.go]: > If not, we should return an error

This makes me modify the cobra `Run` command to `RunE`.

This made me realize that `capAuditor.RunAudits` also returns an error, but we don't actually do anything with it. We probably also should return an error in `RunE`, right ? This feels like a change which is a bit out of scope, so I'll package it in another commit: 0c896e62c2dda9f2f34efb8ec69b794ea8408249

File: internal/capability/auditor.go [hunk 11]
```diff
@@ -26,6 +29,79 @@ type CapAuditor struct {
...
+	if err != nil {
+		logger.Errorf("Error getting absolute path of %s: %v", extraCRDirectory, err)
+	}
+
+	extraCRDirectoryContent, err := os.ReadDir(extraCRDirectoryAbsolutePath)
+	if err != nil {
+		logger.Errorf("Couldn't read %s: %v", extraCRDirectoryAbsolutePath, err)
+	}
+
+	for _, packageDirectory := range extraCRDirectoryContent {
```
[the remark this answers is not in the document]
Comment (mgoerens) [reply internal/capability/auditor.go]: Thanks for that. I'm now using Walk. It simplifies things a bit, though there are other things to check for.

File: internal/capability/auditor.go [hunk 12]
```diff
@@ -26,6 +31,90 @@ type CapAuditor struct {
...
+
+			// Get manifest file content
+			manifestFileContentRaw, err := os.ReadFile(manifestFilePath)
+			if err != nil {
+				logger.Errorf("Error reading file %s: %v", manifestFilePath, err)
+				return nil // continue
+			}
+
+			var manifestFileContentUnmarshalled map[string]interface{}
+			err = yaml.Unmarshal([]byte(manifestFileContentRaw), &manifestFileContentUnmarshalled)
```
Comment (bcrochet) [inline internal/capability/auditor.go:87]: I don't think the []byte() cast is necessary here... os.ReadFile returns []byte

File: internal/capability/auditor.go [hunk 13]
```diff
@@ -26,6 +31,90 @@ type CapAuditor struct {
...
+				return nil // continue
+			}
+
+			// Get the name of the subdirectory containing the manifest file. This corresponds to the package name.
+			packageName := filepath.Base(filepath.Dir(manifestFilePath))
+
+			logger.Debugw("adding Custom Resource", "source manifest file", manifestFilePath, "package", packageName)
+
+			// Get manifest file content
+			manifestFileContentRaw, err := os.ReadFile(manifestFilePath)
```
Comment (bcrochet) [inline internal/capability/auditor.go:80]: Nit: Can we maybe shorten this var to manifestBytes or just manifestRaw?

File: internal/capability/auditor.go [hunk 14]
```diff
@@ -26,6 +31,90 @@ type CapAuditor struct {
...
+			if err != nil {
+				logger.Errorf("Error reading file %s: %v", manifestFilePath, err)
+				return nil // continue
+			}
+
+			var manifestFileContentUnmarshalled map[string]interface{}
```
Comment (bcrochet) [inline internal/capability/auditor.go:86]: Nit: Shorten maybe just manifest?

File: internal/capability/auditor.go [hunk 15]
```diff
@@ -26,6 +31,90 @@ type CapAuditor struct {
 
 	// AllInstallModes will test all install modes supported by an operator
 	AllInstallModes bool
+
+	// ExtraCustomResources associates packages to a list of Custom Resources (in addition to ALMExamples)
+	// to be audited by the OperandInstall AuditPlan.
+	ExtraCustomResources map[string]interface{}
```
Comment (bcrochet) [inline internal/capability/auditor.go:37]: Does this need to be exported?

**Review by bcrochet — CHANGES_REQUESTED** [review]
Couple of nits, a question, and a change.

File: internal/capability/auditor.go (hunk 15 above)
Comment (mgoerens) [reply internal/capability/auditor.go:37]: not anymore

File: internal/capability/auditor.go (hunk 12 above)
Comment (mgoerens) [reply internal/capability/auditor.go:87]: true

File: internal/capability/auditor.go (hunk 13 above)
Comment (mgoerens) [reply internal/capability/auditor.go:80]: I don't have a strong opinion here.
I have a tendency for descriptive (end therefore sometimes long) names, but I guess that can be shorten.

File: internal/capability/auditor.go (hunk 14 above)
Comment (mgoerens) [reply internal/capability/auditor.go:86]: ditto

**Review by bcrochet — APPROVED** [review]
/lgtm

**Review by yashoza19 — APPROVED** [review]
Tested and works as described!

```
yoza@yoza-mac opcap % ./bin/opcap check --packages=cloud-native-postgresql --audit-plan=OperatorInstall,OperandInstall,OperandCleanUp,OperatorCleanUp --log-level=debug --extra-cr-directory=custom_resources_directory/
```

Output:

```

{"package":"cloud-native-postgresql","Operand Kind":"Backup","Operand Name":"backup-sample1","message":"created"}

{"package":"cloud-native-postgresql","Operand Kind":"Backup","Operand Name":"backup-sample","message":"created"}

{"package":"cloud-native-postgresql","Operand Kind":"Cluster","Operand Name":"cluster-sample","message":"created"}

{"package":"cloud-native-postgresql","Operand Kind":"Pooler","Operand Name":"pooler-sample-rw","message":"created"}

{"package":"cloud-native-postgresql","Operand Kind":"ScheduledBackup","Operand Name":"scheduledbackup-sample","message":"created"}
```

**Pull request merged**